### ISSUE TITLE
Support for CQL V2, V1 Binary protocol

### ISSUE THREAD

**Issue by slivne** [scylladb/scylla#584]
Scylla officially support only V3 protocol currently

Need to decide if we are going to support V2, V1 - if not need to enforce this

Delta between V3, V2 (copied from doc/native_protocol_v3.spec)
```
10. Changes from v2
  * stream id is now 2 bytes long (a [short] value), so the header is now 1 byte longer (9 bytes total).
  * BATCH messages now have <flags> (like QUERY and EXECUTE) and a corresponding optional
    <serial_consistency> parameters (see Section 4.1.7).
  * User Defined Types and tuple types have to added to ResultSet metadata (see 4.2.5.2) and a
    new section on the serialization format of UDT and tuple values has been added to the documentation
    (Section 7).
  * The serialization format for collection has changed (both the collection size and
    the length of each argument is now 4 bytes long). See Section 6.
  * QUERY, EXECUTE and BATCH messages can now optionally provide the default timestamp for the query.
    As this feature is optionally enabled by clients, implementing it is at the discretion of the
    client.
  * QUERY, EXECUTE and BATCH messages can now optionally provide the names for the values of the
    query. As this feature is optionally enabled by clients, implementing it is at the discretion of the
    client.
  * The format of "Schema_change" results (Section 4.2.5.5) and "SCHEMA_CHANGE" events (Section 4.2.6)
    has been modified, and now includes changes related to user types.
```

Delta between V2, V1  (copied from doc/native_protocol_v2.spec)

```
9. Changes from v1
  * Protocol is versioned to allow old client connects to a newer server, if a
    newer client connects to an older server, it needs to check if it gets a
    ProtocolException on connection and try connecting with a lower version.
  * A query can now have bind variables even though the statement is not
    prepared; see Section 4.1.4.
  * A new BATCH message allows to batch a set of queries (prepared or not); see 
    Section 4.1.7.
  * Authentication now uses SASL. Concretely, the CREDENTIALS message has been
    removed and replaced by a server/client challenges/responses exchanges (done
    through the new AUTH_RESPONSE/AUTH_CHALLENGE messages). See Section 4.2.3 for
    details.
  * Query paging has been added (Section 7): QUERY and EXECUTE message have an
    additional <result_page_size> [int] and <paging_state> [bytes], and
    the Rows kind of RESULT message has an additional flag and <paging_state> 
    value. Note that paging is optional, and a client that do not want to handle
    can simply avoid including the Page_size flag and parameter in QUERY and
    EXECUTE.
  * QUERY and EXECUTE statements can request for the metadata to be skipped in
    the result set returned (for efficiency reasons) if said metadata are known
    in advance. Furthermore, the result to a PREPARE (section 4.2.5.4) now
    includes the metadata for the result of executing the statement just
    prepared (though those metadata will be empty for non SELECT statements).
```

**Comment by avikivity** [comment]
On 11/18/2015 10:29 AM, slivne wrote:
>
> Scylla officially support only V3 protocol currently
>
> Need to decide if we are going to support V2, V1 - if not need to 
> enforce this
>
> Delta between V3, V2 (copied from doc/native_protocol_v3.spec)
>
> |10. Changes from v2 * stream id is now 2 bytes long (a [short] value), 
> so the header is now 1 byte longer (9 bytes total).|

done

> |* BATCH messages now have <flags> (like QUERY and EXECUTE) and a 
> corresponding optional <serial_consistency> parameters (see Section 
> 4.1.7). * User Defined Types and tuple types have to added to 
> ResultSet metadata (see 4.2.5.2) and a new section on the 
> serialization format of UDT and tuple values has been added to the 
> documentation (Section 7).|

n/a

> |* The serialization format for collection has changed (both the 
> collection size and the length of each argument is now 4 bytes long). 
> See Section 6.|

the biggest issue

> |* QUERY, EXECUTE and BATCH messages can now optionally provide the 
> default timestamp for the query. As this feature is optionally enabled 
> by clients, implementing it is at the discretion of the client.|

done

> |* QUERY, EXECUTE and BATCH messages can now optionally provide the 
> names for the values of the query. As this feature is optionally 
> enabled by clients, implementing it is at the discretion of the client.|

done

> |* The format of "Schema_change" results (Section 4.2.5.5) and 
> "SCHEMA_CHANGE" events (Section 4.2.6) has been modified, and now 
> includes changes related to user types.|

|?

|

**Comment by slivne** [comment]
On Wed, Nov 18, 2015 at 10:31 AM, Avi Kivity <notifications@github.com>
wrote:

>
>
> On 11/18/2015 10:29 AM, slivne wrote:
> >
> > Scylla officially support only V3 protocol currently
> >
> > Need to decide if we are going to support V2, V1 - if not need to
> > enforce this
> >
> > Delta between V3, V2 (copied from doc/native_protocol_v3.spec)
> >
>

What does Done - mean do we support V2 as well ?

> > |10. Changes from v2 * stream id is now 2 bytes long (a [short] value),
> > so the header is now 1 byte longer (9 bytes total).|
>
> done
>

We support V3 but do we need to support also a stream id of 1 BYTE - of V2 ?


>
> > |* BATCH messages now have <flags> (like QUERY and EXECUTE) and a
> > corresponding optional <serial_consistency> parameters (see Section
> > 4.1.7).


We had an issue with that #563 Batch of V2 do not work - so this is not done

**Comment by avikivity** [comment]
On 11/18/2015 10:38 AM, slivne wrote:
> On Wed, Nov 18, 2015 at 10:31 AM, Avi Kivity <notifications@github.com>
> wrote:
>
> >
> >
> > On 11/18/2015 10:29 AM, slivne wrote:
> > >
> > > Scylla officially support only V3 protocol currently
> > >
> > > Need to decide if we are going to support V2, V1 - if not need to
> > > enforce this
> > >
> > > Delta between V3, V2 (copied from doc/native_protocol_v3.spec)
> > >
> >
>
> What does Done - mean do we support V2 as well ?

Done means no work is needed on that item.

>
> > > |10. Changes from v2 * stream id is now 2 bytes long (a [short] 
> value),
> > > so the header is now 1 byte longer (9 bytes total).|
> >
> > done
> >
>
> We support V3 but do we need to support also a stream id of 1 BYTE - 
> of V2 ?
>

It is supported.

>
> >
> > > |* BATCH messages now have <flags> (like QUERY and EXECUTE) and a
> > > corresponding optional <serial_consistency> parameters (see Section
> > > 4.1.7).
>
>
> We had an issue with that #563 Batch of V2 do not work - so this is 
> not done
>

That's why it wasn't marked as done.

**Comment by elcallio** [comment]
As far as I can tell, all the items mentioned above has been addressed. (Batch, collections and schema change response).

**Comment by slivne** [comment]
Calle can ypou please check this issue we had in the past

https://github.com/scylladb/scylla/issues/563

We have moved to use v3 protocol to make sure we are not using v2 for user
profile batching - there at least was an issue when the item was openned

**Comment by elcallio** [comment]
I did. The #563 issue must be at least partially a red herring, because batch parser has always handled different version options (tailing data). Might be that Avi is right and the problem was related to data encoding. Not sure. 
I added a versioned batch dtest as well. (Not that it tests all variants of data, but at least the basics).

**Comment by slivne** [comment]
With regards to #563 can you validate that by running cassandra-stress  from cassandra 2.1.8 and using the profile specified in #563.

**Comment by elcallio** [comment]
Trying to run cassandra-stress from 2.1.8 I get 

NFO  13:23:58 Using data-center name 'datacenter1' for DCAwareRoundRobinPolicy (if this is incorrect, please provide the correct datacenter name with DCAwareRoundRobinPolicy constructor)
INFO  13:23:58 New Cassandra host localhost/127.0.0.1:9042 added
Connected to cluster: Test Cluster
Datatacenter: datacenter1; Host: localhost/127.0.0.1; Rack: rack1
Created schema. Sleeping 1s for propagation.
Sleeping 2s...
Warming up insert with 50000 iterations...
Failed to connect over JMX; not collecting these stats
Generating batches with [2..2] partitions and [0..0] rows (of [2..2] total rows in the partitions)
Exception in thread "main" java.lang.RuntimeException: org.apache.thrift.transport.TTransportException: java.net.ConnectException: Förbindelse vägras
	at org.apache.cassandra.stress.settings.StressSettings.getRawThriftClient(StressSettings.java:144)
	at org.apache.cassandra.stress.settings.StressSettings.getSimpleThriftClient(StressSettings.java:105)
	at org.apache.cassandra.stress.settings.StressSettings.getThriftClient(StressSettings.java:85)
	at org.apache.cassandra.stress.StressProfile.getInsert(StressProfile.java:378)
	at org.apache.cassandra.stress.settings.SettingsCommandUser$1.get(SettingsCommandUser.java:82)
	at org.apache.cassandra.stress.settings.SettingsCommandUser$1.get(SettingsCommandUser.java:78)
	at org.apache.cassandra.stress.operations.SampledOpDistributionFactory$1.get(SampledOpDistributionFactory.java:78)
	at org.apache.cassandra.stress.StressAction$Consumer.<init>(StressAction.java:269)
	at org.apache.cassandra.stress.StressAction.run(StressAction.java:204)
	at org.apache.cassandra.stress.StressAction.warmup(StressAction.java:105)
	at org.apache.cassandra.stress.StressAction.run(StressAction.java:61)
	at org.apache.cassandra.stress.Stress.main(Stress.java:109)
Caused by: org.apache.thrift.transport.TTransportException: java.net.ConnectException: Förbindelse vägras
	at org.apache.thrift.transport.TSocket.open(TSocket.java:187)
	at org.apache.thrift.transport.TFramedTransport.open(TFramedTransport.java:81)
	at org.apache.cassandra.thrift.TFramedTransportFactory.openTransport(TFramedTransportFactory.java:41)
	at org.apache.cassandra.stress.settings.StressSettings.getRawThriftClient(StressSettings.java:124)
	... 11 more
Caused by: java.net.ConnectException: Förbindelse vägras
	at java.net.PlainSocketImpl.socketConnect(Native Method)
	at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:350)
	at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:206)
	at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:188)
	at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392)
	at java.net.Socket.connect(Socket.java:589)
	at org.apache.thrift.transport.TSocket.open(TSocket.java:182)
	... 14 more

Messing with "-mode" (trying native cql3 at what not - not the goal of the exercise btw), does not help. Does cs require thrift to be enabled? Again, counter productive to verifying cql2...

**Comment by slivne** [comment]
Yes ... my bad enable thrift.

Please make sure you use the -user mode this should force using CQL as far
as I recall.

**Comment by elcallio** [comment]
It does not. No combo of -mode settings makes it not using thrift. (Running the exact settings/profile as in 563). And then of course it hits "sorry, not implemented", doing some thrift-thingy...

**Comment by slivne** [comment]
sorrry about it my bad

please checkout scylla-tools-java branch-0.11

compile it - this will provide you the same cassandra-stress tool thatr 563
is using

and run the exact things from 563

I have done that right now

when running the c-s tool from branch-0.11 (with  user
profile=`pwd`/sample.yaml ops\(insert=1\) -rate threads=1
) I get

com.datastax.driver.core.exceptions.DriverInternalError: An unexpected
protocol error occurred. This is a bug in this library, please report:
truncated frame: expected 1 bytes, length is 0
com.datastax.driver.core.exceptions.DriverInternalError: An unexpected
protocol error occurred. This is a bug in this library, please report:
truncated frame: expected 1 bytes, length is 0
com.datastax.driver.core.exceptions.DriverInternalError: An unexpected
protocol error occurred. This is a bug in this library, please report:
truncated frame: expected 1 bytes, length is 0
com.datastax.driver.core.exceptions.DriverInternalError: An unexpected
protocol error occurred. This is a bug in this library, please report:
truncated frame: expected 1 bytes, length is 0
com.datastax.driver.core.exceptions.DriverInternalError: An unexpected
protocol error occurred. This is a bug in this library, please report:
truncated frame: expected 1 bytes, length is 0
com.datastax.driver.core.exceptions.DriverInternalError: An unexpected
protocol error occurred. This is a bug in this library, please report:
truncated frame: expected 1 bytes, length is 0
com.datastax.driver.core.exceptions.DriverInternalError: An unexpected
protocol error occurred. This is a bug in this library, please report:
truncated frame: expected 1 bytes, length is 0
com.datastax.driver.core.exceptions.DriverInternalError: An unexpected
protocol error occurred. This is a bug in this library, please report:
truncated frame: expected 1 bytes, length is 0
com.datastax.driver.core.exceptions.DriverInternalError: An unexpected
protocol error occurred. This is a bug in this library, please report:
truncated frame: expected 1 bytes, length is 0
com.datastax.driver.core.exceptions.DriverInternalError: An unexpected
protocol error occurred. This is a bug in this library, please report:
truncated frame: expected 1 bytes, length is 0
com.datastax.driver.core.exceptions.DriverInternalError: An unexpected
protocol error occurred. This is a bug in this library, please report:
truncated frame: expected 1 bytes, length is 0
com.datastax.driver.core.exceptions.DriverInternalError: An unexpected
protocol error occurred. This is a bug in this library, please report:
truncated frame: expected 1 bytes, length is 0
com.datastax.driver.core.exceptions.DriverInternalError: An unexpected
protocol error occurred. This is a bug in this library, please report:
truncated frame: expected 1 bytes, length is 0
com.datastax.driver.core.exceptions.DriverInternalError: An unexpected
protocol error occurred. This is a bug in this library, please report:
truncated frame: expected 1 bytes, length is 0
com.datastax.driver.core.exceptions.DriverInternalError: An unexpected
protocol error occurred. This is a bug in this library, please report:
truncated frame: expected 1 bytes, length is 0
com.datastax.driver.core.exceptions.DriverInternalError: An unexpected
protocol error occurred. This is a bug in this library, please report:
truncated frame: expected 1 bytes, length is 0
com.datastax.driver.core.exceptions.DriverInternalError: An unexpected
protocol error occurred. This is a bug in this library, please report:
truncated frame: expected 1 bytes, length is 0
com.datastax.driver.core.exceptions.DriverInternalError: An unexpected
protocol error occurred. This is a bug in this library, please report:
truncated frame: expected 1 bytes, length is 0
com.datastax.driver.core.exceptions.DriverInternalError: An unexpected
protocol error occurred. This is a bug in this library, please report:
truncated frame: expected 1 bytes, length is 0
com.datastax.driver.core.exceptions.DriverInternalError: An unexpected
protocol error occurred. This is a bug in this library, please report:
truncated frame: expected 1 bytes, length is 0
com.datastax.driver.core.exceptions.DriverInternalError: An unexpected
protocol error occurred. This is a bug in this library, please report:
truncated frame: expected 1 bytes, length is 0
com.datastax.driver.core.exceptions.DriverInternalError: An unexpected
protocol error occurred. This is a bug in this library, please report:
truncated frame: expected 1 bytes, length is 0
com.datastax.driver.core.exceptions.DriverInternalError: An unexpected
protocol error occurred. This is a bug in this library, please report:
truncated frame: expected 1 bytes, length is 0
com.datastax.driver.core.exceptions.DriverInternalError: An unexpected
protocol error occurred. This is a bug in this library, please report:
truncated frame: expected 1 bytes, length is 0
com.datastax.driver.core.exceptions.DriverInternalError: An unexpected
protocol error occurred. This is a bug in this library, please report:
truncated frame: expected 1 bytes, length is 0


So it seems the issue is not resolved on scylla

using head

commit 1f245e3bcbdaacc5aa2ef139331a2817acae26fb
Author: Avi Kivity <avi@scylladb.com>
Date:   Mon Feb 1 13:34:43 2016 +0200

    mutation_partition: fix use of boost::intrusive::set<>::comp()

    Seems like boost::intrusive::set<>::comp() is not accessible on some
    versions of boost.  Replace by the equivalent
    boost::intrusive::set<>::key_comp().

    Fixes #858.
    Message-Id: <1454326483-29780-1-git-send-email-avi@scylladb.com>

**Comment by elcallio** [comment]
My bad. Ok, I see the problem. A bug in process_batch handling of v2/v3 protocol.
